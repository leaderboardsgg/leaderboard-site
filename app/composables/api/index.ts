export { default as useChangePassword } from './useChangePassword'
export { default as useConfirmAccount } from './useConfirmAccount'
export { default as useGetLeaderboardBySlug } from './useGetLeaderboardBySlug'
export { default as useGetLeaderboards } from './useGetLeaderboards'
export { default as useGetUser } from './useGetUser'
export { default as useLoginUser } from './useLoginUser'
export { default as useLogoutUser } from './useLogoutUser'
export { default as useRecoverAccount } from './useRecoverAccount'
export { default as useRegisterUser } from './useRegisterUser'
export { default as useValidateRecoveryToken } from './useValidateRecoveryToken'
export { default as useResendAccountConfirmation } from './useResendAccountConfirmation'
