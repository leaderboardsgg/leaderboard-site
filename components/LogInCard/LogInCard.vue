<template>
  <BaseCard id="loginCard" class="bg-white w-full max-w-xl">
    <BaseCardHeader class="flex flex-row space-x-3">
      <div
        class="
          flex flex-1
          justify-center
          px-3
          py-2
          rounded
          bg-gray-100
          text-gray-900
        "
      >
        Log In
      </div>

      <BaseButton
        class="
          flex flex-1
          justify-center
          bg-white
          text-gray-900
          border border-gray-300
          hover:bg-gray-100
        "
        @click="$emit('signUpClick')"
      >
        Sign Up
      </BaseButton>

      <CoreCloseButton v-show="modal" @click.prevent="$emit('close')" />
    </BaseCardHeader>

    <BaseCardBody>
      <div class="flex flex-col space-y-3 pb-3 mb-3 border-b border-gray-300">
        <BaseInput
          v-model="email"
          class="bg-white"
          name="email"
          type="text"
          placeholder="Email"
          autocomplete="email"
        />

        <div class="relative flex flex-col w-full">
          <BaseInput
            v-model="password"
            class="bg-white pr-12"
            name="password"
            :type="showPassword ? 'text' : 'password'"
            placeholder="Password"
            autocomplete="password"
          />

          <div class="absolute top-0 right-0 flex items-center h-full">
            <BaseButton
              type="submit"
              class="
                p-1
                mr-2
                rounded
                fill-current
                text-gray-500
                hover:bg-gray-100 hover:text-gray-600
              "
              @click="showPassword = !showPassword"
            >
              <svg
                v-show="!showPassword"
                xmlns="http://www.w3.org/2000/svg"
                class="h-5 w-5"
                viewBox="0 0 20 20"
                fill="currentColor"
              >
                <path
                  fill-rule="evenodd"
                  d="M3.707 2.293a1 1 0 00-1.414 1.414l14 14a1 1 0 001.414-1.414l-1.473-1.473A10.014 10.014 0 0019.542 10C18.268 5.943 14.478 3 10 3a9.958 9.958 0 00-4.512 1.074l-1.78-1.781zm4.261 4.26l1.514 1.515a2.003 2.003 0 012.45 2.45l1.514 1.514a4 4 0 00-5.478-5.478z"
                  clip-rule="evenodd"
                />
                <path
                  d="M12.454 16.697L9.75 13.992a4 4 0 01-3.742-3.741L2.335 6.578A9.98 9.98 0 00.458 10c1.274 4.057 5.065 7 9.542 7 .847 0 1.669-.105 2.454-.303z"
                />
              </svg>

              <svg
                v-show="showPassword"
                xmlns="http://www.w3.org/2000/svg"
                class="h-5 w-5"
                viewBox="0 0 20 20"
                fill="currentColor"
              >
                <path d="M10 12a2 2 0 100-4 2 2 0 000 4z" />
                <path
                  fill-rule="evenodd"
                  d="M.458 10C1.732 5.943 5.522 3 10 3s8.268 2.943 9.542 7c-1.274 4.057-5.064 7-9.542 7S1.732 14.057.458 10zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
                  clip-rule="evenodd"
                />
              </svg>
            </BaseButton>
          </div>
        </div>

        <BaseButton to="#" class="login-button">Log In</BaseButton>
      </div>

      <div class="flex flex-col w-full space-y-2">
        <BaseButton class="login-button">
          <svg
            class="w-5 h-5 mr-2"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 480 512"
          >
            <path
              fill="currentColor"
              d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"
            ></path>
          </svg>

          <p>Log In with Github</p>
        </BaseButton>

        <BaseButton class="login-button">
          <svg
            class="w-5 h-5 mr-2"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 488 512"
          >
            <path
              fill="currentColor"
              d="M488 261.8C488 403.3 391.1 504 248 504 110.8 504 0 393.2 0 256S110.8 8 248 8c66.8 0 123 24.5 166.3 64.9l-67.5 64.9C258.5 52.6 94.3 116.6 94.3 256c0 86.5 69.1 156.6 153.7 156.6 98.2 0 135-70.4 140.8-106.9H248v-85.3h236.1c2.3 12.7 3.9 24.9 3.9 41.4z"
            ></path>
          </svg>

          <p>Log In with Google</p>
        </BaseButton>
      </div>
    </BaseCardBody>
  </BaseCard>
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs } from '@nuxtjs/composition-api';

export default defineComponent({
  props: {
    modal: {
      type: Boolean,
      required: false,
      default: false,
    },
  },

  setup() {
    const state = reactive({
      email: '',
      password: '',
      showPassword: false,
    });

    return {
      ...toRefs(state),
    };
  },
});
</script>

<style scoped>
.login-button {
  @apply flex flex-1 items-center justify-center fill-current bg-gray-100 text-gray-900 hover:bg-gray-200;
}
</style>
