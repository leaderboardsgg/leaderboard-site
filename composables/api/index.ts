export { default as useLoginUser } from './useLoginUser'
export { default as useLogoutUser } from './useLogoutUser'
export { default as useRegisterUser } from './useRegisterUser'
