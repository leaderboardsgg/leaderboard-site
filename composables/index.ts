export { default as useCurrentUser } from './useCurrentUser'
export { default as useSessionToken } from './useSessionToken'
