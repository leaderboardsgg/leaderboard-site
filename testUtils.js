import Vue from 'vue';
import { RouterLinkStub } from '@vue/test-utils';
import userEvent from '@testing-library/user-event';
import { render } from '@testing-library/vue';

export const NuxtLinkStub = Vue.component('NuxtLinkStub', {
  extends: RouterLinkStub,
  render(h) {
    const temp = h(
      this.tag || 'a',
      {
        attrs: {
          ...this.$vnode.data.attrs,
          ...this.$attrs,
          href: this.to,
        },
      },
      this.$slots.default,
    );

    return temp;
  },
});

export const stubbedRender = (Component, options) => {
  return render(Component, {
    stubs: { NuxtLink: NuxtLinkStub, ...options?.stubs },
    ...options,
  });
};

export * from '@testing-library/vue';
export * from '@vue/test-utils';
export { userEvent as fireEvent };
