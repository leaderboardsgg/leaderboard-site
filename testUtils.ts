import { createElementBlock, Component, ComponentOptions } from 'vue'
import { MountingOptions } from '@vue/test-utils'
import userEvent from '@testing-library/user-event'
import { render, RenderOptions } from '@testing-library/vue'

export const stubbedRender = (
  Component: Component | ComponentOptions,
  options?: RenderOptions | MountingOptions<any>,
) => {
  return render(Component, {
    global: {
      stubs: {
        NuxtLink: createElementBlock('a'),
      },
    },
    ...options,
  })
}

export * from '@testing-library/vue'
export * from '@nuxt/test-utils'
export * from '@vue/test-utils'
export { userEvent as fireEvent }
