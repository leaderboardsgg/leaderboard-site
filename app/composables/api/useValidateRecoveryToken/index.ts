import { useRuntimeConfig } from '#app'
import {
  useApi,
  type ApiResponse,
  type optionalParameters,
} from 'composables/useApi'
import { Account } from 'lib/api/Account'

export async function useValidateRecoveryToken(
  recoveryToken: string,
  opts: optionalParameters<void> = {},
): Promise<ApiResponse<void>> {
  const { onError, onOkay } = opts

  const account = new Account({
    baseUrl: useRuntimeConfig().public.backendBaseUrl,
  })

  // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
  return await useApi<void>(
    async () => await account.testRecoveryToken(recoveryToken),
    { onError, onOkay },
  )
}

export default useValidateRecoveryToken
