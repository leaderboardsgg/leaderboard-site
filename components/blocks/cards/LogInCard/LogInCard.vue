<script setup lang="ts">
import type { Ref } from 'vue'
import BaseButton from 'elements/buttons/BaseButton/BaseButton.vue'
import CloseButton from 'elements/buttons/CloseButton/CloseButton.vue'
import BaseInput from 'elements/inputs/BaseInput/BaseInput.vue'
import HideShowPassword from 'elements/buttons/HideShowPassword/HideShowPassword.vue'
import CardBody from 'elements/cards/CardBody/CardBody.vue'
import CardHeader from 'elements/cards/CardHeader/CardHeader.vue'
import Card from 'elements/cards/Card/Card.vue'
import { useLoginUser } from 'composables/api/useLoginUser'

interface LogInCardProps {
  modal?: boolean
}

interface LogInCardState {
  email: Ref<string>
  password: Ref<string>
  showPassword: Ref<boolean>
}

const emit = defineEmits<{
  (event: 'close'): void
  (event: 'signUpClick'): void
}>()

const props = withDefaults(defineProps<LogInCardProps>(), {
  modal: false,
})

const state: LogInCardState = {
  email: ref(''),
  password: ref(''),
  showPassword: ref(false),
}

function login() {
  useLoginUser({
    email: state.email.value,
    password: state.password.value,
  })

  state.email.value = ''
  state.password.value = ''
  state.showPassword.value = false

  emit('close')
}
</script>

<template>
  <Card id="logInCard" data-testid="log-in-card" class="login-card">
    <CardHeader class="login-card__header">
      <div class="login-card__title">Log In</div>

      <BaseButton
        id="sign-up-button"
        class="login-card__sign-up-button"
        data-testid="sign-up-button"
        @click="emit('signUpClick')"
      >
        Sign Up
      </BaseButton>

      <CloseButton
        v-show="props.modal"
        data-testid="close-button"
        @click.prevent="emit('close')"
      />
    </CardHeader>

    <CardBody>
      <div class="login-card__body-wrapper">
        <BaseInput
          :model="state.email"
          name="email"
          type="text"
          placeholder="Email"
          autocomplete="email"
          data-testid="email-input"
        />

        <div class="login-card__input-wrapper">
          <BaseInput
            :model="state.password"
            class="login-card__password-field"
            name="password"
            :type="state.showPassword.value ? 'text' : 'password'"
            placeholder="Password"
            autocomplete="password"
            data-testid="password-input"
            @keyup.enter="login"
          />

          <div class="login-card__button-wrapper">
            <HideShowPassword
              id="hide-show-button"
              type="button"
              data-testid="hide-show-button"
              @click="state.showPassword.value = !state.showPassword.value"
              @keydown.enter.prevent=""
              @keyup.enter="
                state.showPassword.value = !state.showPassword.value
              "
              @keydown.enter="$event.preventDefault()"
            />
          </div>
        </div>

        <BaseButton
          class="login-button"
          data-testid="login-button"
          @click="login"
        >
          Log In
        </BaseButton>
      </div>

      <div class="login-card__auth-buttons">
        <BaseButton class="login-button">
          <i-svg-github class="mr-2 h-5 w-5" />

          <p>Log In with Github</p>
        </BaseButton>

        <BaseButton class="login-button">
          <i-svg-google class="mr-2 h-5 w-5" />
          <p>Log In with Google</p>
        </BaseButton>
      </div>
    </CardBody>
  </Card>
</template>

<style lang="postcss" scoped>
.login-card {
  @apply bg-white w-full max-w-xl;
  & .login-card__header {
    @apply flex flex-row space-x-3;
  }
  & .login-card__title {
    @apply flex flex-1 justify-center px-3 py-2 rounded bg-gray-100 text-gray-900;
  }
  & .login-card__sign-up-button {
    @apply flex flex-1 justify-center bg-white text-gray-900 border border-gray-300;
  }
  & .login-card__body-wrapper {
    @apply flex flex-col space-y-3 pb-3 mb-3 border-b border-gray-300;
  }
  & .login-card__input-wrapper {
    @apply relative flex flex-col w-full;
  }
  & .login-card__button-wrapper {
    @apply absolute top-0 right-0 flex items-center h-full;
  }
  & .login-card__auth-buttons {
    @apply flex flex-col w-full space-y-2;
  }
  & .login-card__password-field {
    @apply bg-white pr-12;
  }
}

#hide-show-button {
  @apply text-gray-700 hover:bg-gray-100;
  @apply p-1 mr-2 rounded fill-current;
}

#sign-up-button {
  @apply hover:bg-gray-100;
}

.login-button {
  @apply flex flex-1 items-center justify-center fill-current bg-gray-100 text-gray-900 hover:bg-gray-200;
}
</style>
