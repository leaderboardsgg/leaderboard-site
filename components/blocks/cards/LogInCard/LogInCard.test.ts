import { mount, enableAutoUnmount } from '@vue/test-utils'
import { getByTestId, getHTMLElement } from 'root/testUtils'
import * as apiComposables from 'composables/api'
import LogInCard from './LogInCard.vue'

const token = 'jwt-token'
const mockSuccessUsersLoginCreate = vi.fn(() =>
  Promise.resolve({ data: { token }, ok: true }),
)
const mockSuccessUsersMeList = vi.fn(() =>
  Promise.resolve({
    data: { id: 1, username: 'admin' },
    ok: true,
  }),
)

function getLogInCardWrapper() {
  return mount(LogInCard)
}

afterEach(() => {
  vi.restoreAllMocks()
})

enableAutoUnmount(afterEach)

describe('<LogInCard />', () => {
  beforeEach(() => {
    vi.mock('lib/api/Users', () => ({
      Users: function Users() {
        this.usersLoginCreate = mockSuccessUsersLoginCreate
        this.usersMeList = mockSuccessUsersMeList
      },
    }))
  })

  it('should render without crashing', () => {
    const wrapper = getLogInCardWrapper()

    expect(wrapper.isVisible()).toBe(true)
  })

  describe('when the close button is clicked', () => {
    it('should emit the close event', async () => {
      const wrapper = getLogInCardWrapper()

      const closeButton = getByTestId(wrapper, 'close-button')

      await closeButton.trigger('click')

      expect(wrapper.emitted().close).toBeTruthy()
    })
  })

  describe('when the hide/show button is clicked', () => {
    it('changes the password input type to be text', async () => {
      const wrapper = getLogInCardWrapper()
      const passwordInputElement = getHTMLElement(
        getByTestId(wrapper, 'password-input'),
      ) as HTMLInputElement

      expect(passwordInputElement.type).toBe('password')

      await getByTestId(wrapper, 'hide-show-button').trigger('click')

      expect(passwordInputElement.type).toBe('text')
    })
  })

  // TODO: skip this for now
  describe.skip('when enter key is released on the password input field', () => {
    it('emits the close event', async () => {
      const wrapper = getLogInCardWrapper()

      await getByTestId(wrapper, 'password-input').trigger('keyup.enter')

      expect(wrapper.emitted().close).toBeTruthy()
    })
  })

  // TODO: skip this for now
  describe.skip('when the login button is clicked', () => {
    const emailAddress = 'strongbad@homestarrunner.com'
    const password = 'homestarsux'

    it('emits the close event', async () => {
      const wrapper = getLogInCardWrapper()

      const emailInput = getByTestId(wrapper, 'email-input')
      const passwordInput = getByTestId(wrapper, 'password-input')

      await emailInput.setValue(emailAddress)
      await passwordInput.setValue(password)

      await getByTestId(wrapper, 'login-button').trigger('click')

      expect(wrapper.emitted().close).toBeTruthy()
    })

    it('clears the state', async () => {
      const wrapper = getLogInCardWrapper()

      const emailInput = getByTestId(wrapper, 'email-input')
      const passwordInput = getByTestId(wrapper, 'password-input')

      await emailInput.setValue(emailAddress)
      await passwordInput.setValue(password)

      const emailInputElement = getHTMLElement(emailInput) as HTMLInputElement
      const passwordInputElement = getHTMLElement(
        passwordInput,
      ) as HTMLInputElement

      expect(emailInputElement.value).toBe(emailAddress)
      expect(passwordInputElement.value).toBe(password)

      await getByTestId(wrapper, 'login-button').trigger('click')

      expect(emailInputElement.value).toBe('')
      expect(passwordInputElement.value).toBe('')
    })

    // this test is still failing
    it('calls the api', async () => {
      const useLoginUserSpy = vi.spyOn(apiComposables, 'useLoginUser')

      const wrapper = getLogInCardWrapper()

      const emailInput = getByTestId(wrapper, 'email-input')
      const passwordInput = getByTestId(wrapper, 'password-input')

      await emailInput.setValue(emailAddress)
      await passwordInput.setValue(password)

      await getByTestId(wrapper, 'login-button').trigger('click')

      expect(useLoginUserSpy).toBeCalledTimes(1)
    })
  })

  describe('when the sign up button is clicked', () => {
    it('emits the sign up click event', async () => {
      const wrapper = getLogInCardWrapper()

      await getByTestId(wrapper, 'sign-up-button').trigger('click')

      expect(wrapper.emitted().signUpClick).toBeTruthy()
    })
  })
})
