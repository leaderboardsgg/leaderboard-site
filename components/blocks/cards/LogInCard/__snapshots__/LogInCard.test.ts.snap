// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LogInCard /> renders correctly 1`] = `
<div
  class="card login-card"
  data-testid="login-card"
  id="loginCard"
>
  <div
    class="card-header login-card__header"
  >
    <div
      class="login-card__title"
    >
      Log In
    </div>
     
    <button
      class="button login-card__sign-up-button"
      id="sign-up-button"
    >
      
      Sign Up
    
    </button>
     
    <button
      class="button close-button"
      style="display: none;"
    >
      <svg-icon-stub
        class="w-5 h-5"
        name="close"
      />
    </button>
  </div>
   
  <div
    class="card-body"
  >
    <div
      class="login-card__body-wrapper"
    >
      <input
        autocomplete="email"
        class="input"
        name="email"
        placeholder="Email"
        type="text"
      />
       
      <div
        class="login-card__input-wrapper"
      >
        <input
          autocomplete="password"
          class="input login-card__password-field"
          name="password"
          placeholder="Password"
          type="password"
        />
         
        <div
          class="login-card__button-wrapper"
        >
          <button
            class="button button"
            data-testid="hide-show-password-button"
            id="hide-show-button"
            type="button"
          >
            <svg-icon-stub
              class="w-5 h-5"
              data-testid="hidden-eye-icon"
              name="eye-hidden"
            />
             
            <svg-icon-stub
              class="w-5 h-5"
              data-testid="visible-eye-icon"
              name="eye-visible"
              style="display: none;"
            />
          </button>
        </div>
      </div>
       
      <button
        class="button login-button"
      >
         Log In 
      </button>
    </div>
     
    <div
      class="login-card__auth-buttons"
    >
      <button
        class="button login-button"
      >
        <svg-icon-stub
          class="mr-2 w-5 h-5"
          name="github"
        />
         
        <p>
          Log In with Github
        </p>
      </button>
       
      <button
        class="button login-button"
      >
        <svg-icon-stub
          class="mr-2 w-5 h-5"
          name="google"
        />
         
        <p>
          Log In with Google
        </p>
      </button>
    </div>
  </div>
</div>
`;
